File: code/greedy_algorithms/src/minimum_coins/minimum_coins.go
package main

import (
	"fmt"
)

type Scenario struct {
	Value  int
	Denoms []int
	Result []int
}

func main() {
	scenarios := []Scenario{
		Scenario{Value: 100, Denoms: []int{50, 25, 10, 5, 1}, Result: []int{50, 50}},
		Scenario{Value: 101, Denoms: []int{50, 25, 10, 5, 1}, Result: []int{50, 50, 1}},
		Scenario{Value: 77, Denoms: []int{50, 25, 10, 5, 1}, Result: []int{50, 25, 1, 1}},
		Scenario{Value: 38, Denoms: []int{50, 25, 10, 5, 1}, Result: []int{25, 10, 1, 1, 1}},
		Scenario{Value: 17, Denoms: []int{50, 25, 10, 5, 1}, Result: []int{10, 5, 1, 1}},
		Scenario{Value: 3, Denoms: []int{50, 25, 10, 5, 1}, Result: []int{1, 1, 1}},
		Scenario{Value: 191, Denoms: []int{100, 50, 25, 10, 5, 1}, Result: []int{100, 50, 25, 10, 5, 1}},
	}

	for _, s := range scenarios {
		result := getMinimumCoins(s.Value, s.Denoms)
		if fmt.Sprint(result) != fmt.Sprint(s.Result) {
			fmt.Println("Test Failed: Value: ", s.Value, ", Denominations: ", s.Denoms, ", Expected Result: ", s.Result, ", Actual Result: ", result)
		}
	}
}

func getMinimumCoins(value int, denoms []int) []int {
	var result []int
	for _, d := range denoms {
		for {
			if value < d {
				break
			}
			result = append(result, d)
			value -= d
		}
	}
	return result
}

File: cosmos-master/code/greedy-algorithms/minimum_coins/minimum-coins.js
/* Part of Cosmos by OpenGenus Foundation */

function minimumCoins(value, denominations) {
    var result = [];
    // Assuming denominations is sorted in descendig order
    for(var i = 0; i < denominations.length; i++) {
        var cur_denom = denominations[i];
        while(cur_denom <= value) {
            result.push(cur_denom);
            value -= cur_denom;
        }
    }
    return result;
}

Array.prototype.equals = function(other) {
    if (!other || !(other instanceof Array)) {
        return false;
    }

    if (this.length != other.length) {
        return false;
    }

    for (var i = 0; i < this.length; i++) {
        if (this[i] != other[i]) {
            return false;
        }
    }

    return true;
}

function test() {
    scenarios = [
        {value: 100, denoms: [50, 25, 10, 5, 1],      result: [50, 50]},
        {value: 101, denoms: [50, 25, 10, 5, 1],      result: [50, 50, 1]},
        {value: 77,  denoms: [50, 25, 10, 5, 1],      result: [50, 25, 1, 1]},
        {value: 38,  denoms: [50, 25, 10, 5, 1],      result: [25, 10, 1, 1, 1]},
        {value: 17,  denoms: [50, 25, 10, 5, 1],      result: [10, 5, 1, 1]},
        {value: 3,   denoms: [50, 25, 10, 5, 1],      result: [1, 1, 1]},
        {value: 191, denoms: [100, 50, 25, 10, 5, 1], result: [100, 50, 25, 10, 5, 1]}
    ];

    scenarios.forEach(function(scenario) {
        var actual = minimumCoins(scenario.value, scenario.denoms);
        if (!scenario.result.equals(actual)) {
            console.error("Test Failed: Value: " + scenario.value
                        + ", Denominations: " + scenario.denoms
                        + ", Expected Result: " + scenario.result
                        + ", Actual Result: " + actual);
        }
    }, this);
}

test();

